docs/proposals-accepted/202205-vertical-query-sharding.md

```diff
@@ -0,0 +1,152 @@
+
+# hash(pod=prometheus-2) mod 2 = 14949237384223363101 mod 2 = 1
+memory_usage_bytes{pod="prometheus-2", region="us-1", role="apps"}
+```
+
+The first two series would therefore end up in the first query shard, and the second two series on the second query shard. The Queriers will execute the query for their own subset only, and the sharding component will concatenate the results before returning them to the users.
+
+The reason why partitioning the series set on the grouping labels works is because a grouping aggregation can be executed independently on each combination of values for the labels found in the grouping clause. In our example, as long as series with the same `pod` label end up in the same query shard, we can safely perform the shard and merge strategy. 
+
+### Initiating sharded queries
+
+The Query Frontend component already has useful splitting and merging capabilities which are currently used for sharding queries horizontally. In order to provide the best possible user experience and not burden users with running additional component(s), we propose adding a new middleware to Query Frontend which will implement the vertical sharding algorithm proposed here after the step-alignment and horizontal time-slicing steps. This would allow users to restrict the maximum complexity of a given query based on vertical sharding of already time-sliced queries. The only new user-specified parameter would be the number of shards in which to split PromQL aggregations.
+
+Integrating vertical-query sharding in QFE also has the added benefits of: 
+* Utilizing the existing query range caching implementation 
+* Vertical-query sharding a query that has already been time-sliced reduces the cardinality of each respective query shard
+The following diagram illustrates how the sharding algorithm would work end to end:
+
+![Vertical sharding](../img/vertical-sharding.png "Vertical query sharding")
+
+### Drawbacks & Exceptions
+
+*Not all queries are easily shardeable*
+There are certain aggregations for which sharding cannot be performed safely. For these cases the sharder can simply fall back to completely executing the expression as a single query. These cases include the use of functions such as `label_replace` and `label_join` which create new labels inside PromQL, when the query is being executed. Since such labels can be arbitrarily created while executing queries, Stores will be unaware of them and cannot take them into account when sharding the matching series set. 
+
+#TODO Mention promql parsing/transformation to increase queries that can be sharded
+
+*Impact of sharded queries on block retrieval* 
+Given that queries are sharded based on series, and not time, for a given leaf most metrics for each independent shard will likely reside in the same block(s). This can multiply the cost of retrieving series from a given block since the block has to be traversed N number of times, where N is the number of distinct shards.
```

Comment:
`groupcache` should alleviate these problems ;P

Reply:
With block-sharded store-gw, will store-gw in the same groupcache ever need to talk to each other?